### PR TITLE
Chore: Upgrade Go to 1.21.3

### PR DESCRIPTION
**Pull request by ryantxu**
**What is this feature?**

Update the min go version to 1.21. (see also #76359)

**Why do we need this feature?**

This lets us use new features added in 1.21.  This lets us use the new standard library functions rather than the ones in `"golang.org/x/exp/..." (eg slices+maps)

**Who is this feature for?**

Developers

### CODE REVIEW

File: pkg/web/macaron.go [hunk 1]
```diff
@@ -1,6 +1,3 @@
-//go:build go1.3
-// +build go1.3
```
Comment (ryantxu) [inline pkg/web/macaron.go:2]: with these tags, the build results are:
```
go build -ldflags -w -X main.version=10.3.0-pre -X main.commit=caa185df53 -X main.buildstamp=1698422981 -X main.buildBranch=use-golant-121 -o ./bin/grafana ./pkg/cmd/grafana
# github.com/grafana/grafana/pkg/web
pkg/web/macaron.go:48:14: predeclared any requires go1.18 or later (-lang was set to go1.21; check go.mod)
pkg/web/macaron.go:73:13: cannot use http.NotFound (value of type func(w http.ResponseWriter, r *http.Request)) as Handler value in argument to m.NotFound
pkg/web/macaron.go:107:6: type aliases requires go1.9 or later (-lang was set to go1.21; check go.mod)
pkg/web/macaron.go:122:15: invalid operation: handler (variable of type Handler) is not an interface
exit status 1
exit status 1
make[1]: *** [build-go] Error 1
```

File: go.mod [hunk 2]
```diff
@@ -1,6 +1,6 @@
 module github.com/grafana/grafana
 
-go 1.20
+go 1.21
```
Comment (kminehart) [inline go.mod:3]: you should leave this unless you're using specific features from 1.21

Comment (ryantxu) [reply go.mod:3]: ya -- the hope is to use the 1.21 features :)    it moved a bunch of things from `golang.org/x/exp/...` to the standard import path

Comment (kminehart) [reply go.mod:3]: nicee

File: .github/CODEOWNERS [hunk 3]
```diff
@@ -635,7 +635,6 @@ embed.go @grafana/grafana-as-code
 /.github/workflows/issue-opened.yml @grafana/grafana-community-support
 /.github/workflows/metrics-collector.yml @torkelo
 /.github/workflows/milestone.yml @marefr
-/.github/workflows/ox-code-coverage.yml @grafana/explore-squad
```
Comment (ryantxu) [inline .github/CODEOWNERS:638]: removed since not used by explore anymore... see https://github.com/grafana/code-coverage/pull/29#issuecomment-1788612302

**Review by kminehart — APPROVED** [review]

File: .github/CODEOWNERS (hunk 3 above)
Comment (matyax) [reply .github/CODEOWNERS:638]: The @grafana/observability-logs was using that workflow to report coverage changes btw.

**Pull request merged**